File: services/ai-orchestrator/app/providers/openai.py
"""OpenAI adapter — structured output enforced via JSON Schema response_format."""
from __future__ import annotations

import json
from typing import List, Optional

import httpx

from ..schemas import CONTRIBUTION_JSON_SCHEMA, Category, StructuredContribution
from .base import ProviderError

SYSTEM_PROMPT = """You are the contribution parser of a living-planet simulation.
Convert the user's free-text idea (Arabic or English) into a structured game
element. All trait values must be floats in [0,1]. Keep the element balanced:
no immortality, no infinite energy, no unbounded reproduction. Infer the most
suitable biomes and realistic ecological/economic risks. Output JSON only."""


class OpenAIProvider:
    name = "openai"

    def __init__(self, api_key: str, model: str = "gpt-4o-mini", base_url: str = "https://api.openai.com"):
        self.api_key = api_key
        self.model = model
        self.base_url = base_url

    async def parse_contribution(self, text: str, category: Optional[Category], locale: str) -> StructuredContribution:
        payload = {
            "model": self.model,
            "messages": [
                {"role": "system", "content": SYSTEM_PROMPT},
                {"role": "user", "content": f"locale={locale}; category_hint={category.value if category else 'auto'}\nidea: {text}"},
            ],
            "response_format": {
                "type": "json_schema",
                "json_schema": {
                    "name": "structured_contribution",
                    "strict": True,
                    "schema": CONTRIBUTION_JSON_SCHEMA,
                },
            },
            "temperature": 0.2,
        }
        async with httpx.AsyncClient(timeout=25) as client:
            response = await client.post(
                f"{self.base_url}/v1/chat/completions",
                headers={"authorization": f"Bearer {self.api_key}"},
                json=payload,
            )
        if response.status_code != 200:
            raise ProviderError(f"openai {response.status_code}: {response.text[:200]}")
        data = response.json()
        try:
            content = data["choices"][0]["message"]["content"]
            parsed = json.loads(content)
            # model may wrap in an object keyed by the schema name
            if "structured_contribution" in parsed:
                parsed = parsed["structured_contribution"]
            return StructuredContribution.model_validate(parsed)
        except (KeyError, IndexError, json.JSONDecodeError) as exc:
            raise ProviderError(f"openai malformed response: {exc}") from exc

    async def narrate(self, contribution_name: str, facts: List[str], locale: str) -> str:
        payload = {
            "model": self.model,
            "messages": [
                {"role": "system", "content": (
                    "You narrate simulation results. You may ONLY mention the facts "
                    "listed by the user; inventing events is forbidden. "
                    + ("Write in Arabic." if locale == "ar" else "Write in English.")
                )},
                {"role": "user", "content": f"element: {contribution_name}\nfacts:\n" + "\n".join(facts)},
            ],
            "temperature": 0.6,
        }
        async with httpx.AsyncClient(timeout=25) as client:
            response = await client.post(
                f"{self.base_url}/v1/chat/completions",
                headers={"authorization": f"Bearer {self.api_key}"},
                json=payload,
            )
        if response.status_code != 200:
            raise ProviderError(f"openai {response.status_code}")
        return response.json()["choices"][0]["message"]["content"].strip()

File: planet/services/ai-orchestrator/src/providers/openai.ts
import { ContributionCategory, type AnalyzedContribution } from "@planet/shared-types";
import { ProviderUnavailableError, type AnalyzeRequest, type LLMProvider, type NarrateRequest, type NarrateResult } from "./base.js";

const ANALYZE_SCHEMA = {
  type: "object",
  properties: {
    name: { type: "string" },
    summary: { type: "string" },
    traits: {
      type: "object",
      properties: Object.fromEntries(
        ["size", "growthRate", "waterNeed", "energyOutput", "pollutionAbsorption", "pollutionEmission", "coldResistance", "heatResistance", "aggression", "intelligence", "reproductionRate", "lifespan", "adaptability", "rarity", "nightLuminosity", "toxicity", "spreadRate"].map((k) => [k, { type: "number", minimum: 0, maximum: 1 }]),
      ),
      required: ["size", "growthRate", "waterNeed", "energyOutput", "pollutionAbsorption", "pollutionEmission", "coldResistance", "heatResistance", "aggression", "intelligence", "reproductionRate", "lifespan", "adaptability", "rarity", "nightLuminosity", "toxicity", "spreadRate"],
      additionalProperties: false,
    },
    possibleBiomes: { type: "array", items: { type: "string" } },
    risks: { type: "array", items: { type: "string" } },
    benefits: { type: "array", items: { type: "string" } },
  },
  required: ["name", "summary", "traits", "possibleBiomes", "risks", "benefits"],
  additionalProperties: false,
};

/** OpenAI adapter — live only when OPENAI_API_KEY is configured. */
export class OpenAIProvider implements LLMProvider {
  name = "openai";
  private apiKey = process.env["OPENAI_API_KEY"] ?? "";
  private model = process.env["OPENAI_MODEL"] ?? "gpt-4o-mini";

  isLive(): boolean {
    return this.apiKey.length > 0;
  }

  async analyze(req: AnalyzeRequest): Promise<AnalyzedContribution> {
    if (!this.isLive()) throw new ProviderUnavailableError(this.name);
    const res = await fetch("https://api.openai.com/v1/chat/completions", {
      method: "POST",
      headers: { "content-type": "application/json", authorization: `Bearer ${this.apiKey}` },
      body: JSON.stringify({
        model: this.model,
        response_format: {
          type: "json_schema",
          json_schema: { name: "contribution_analysis", schema: ANALYZE_SCHEMA, strict: true },
        },
        messages: [
          { role: "system", content: ANALYSIS_SYSTEM_PROMPT },
          { role: "user", content: `category: ${req.category}\ntext: ${req.text}` },
        ],
        temperature: 0.2,
      }),
    });
    if (!res.ok) throw new Error(`openai_http_${res.status}`);
    const data = (await res.json()) as { choices: { message: { content: string } }[] };
    const parsed = JSON.parse(data.choices[0]?.message.content ?? "{}") as Omit<AnalyzedContribution, "category" | "sandbox">;
    return { ...parsed, category: req.category, sandbox: false };
  }

  async narrate(req: NarrateRequest): Promise<NarrateResult> {
    if (!this.isLive()) throw new ProviderUnavailableError(this.name);
    const res = await fetch("https://api.openai.com/v1/chat/completions", {
      method: "POST",
      headers: { "content-type": "application/json", authorization: `Bearer ${this.apiKey}` },
      body: JSON.stringify({
        model: this.model,
        messages: [
          { role: "system", content: NARRATIVE_SYSTEM_PROMPT(req.locale) },
          { role: "user", content: JSON.stringify(req.facts) },
        ],
        temperature: 0.7,
      }),
    });
    if (!res.ok) throw new Error(`openai_http_${res.status}`);
    const data = (await res.json()) as { choices: { message: { content: string } }[] };
    return { text: data.choices[0]?.message.content ?? "", llmGenerated: true };
  }
}

export const ANALYSIS_SYSTEM_PROMPT = `You convert a user's free-form world element idea into normalized simulation traits (0..1). Be conservative and balanced: nothing is immortal, infinite, or omnipotent. Infer biomes, risks and benefits from the description. Output only the JSON schema fields.`;

export const NARRATIVE_SYSTEM_PROMPT = (locale: "ar" | "en") =>
  locale === "ar"
    ? `أنت مؤرخ عالم محاكاة. اصغ قصة قصيرة مترابطة من قائمة الحقائق فقط. ممنوع منعاً باتاً ذكر أي حدث أو رقم غير موجود في الحقائق. اكتب بالعربية الفصحى.`
    : `You are the historian of a simulated world. Weave a short coherent story from the provided facts only. Never invent events or numbers that are not in the facts. Write in English.`;

export { ContributionCategory };
